### PR TITLE
Fix web editable text composing range

### PR DESCRIPTION
**Pull request by antholeole**
Flutter framework on web was not getting `composingRange` updates from the engine. This is the fix. Loosely based on [this PR](https://github.com/flutter/engine/pull/25673). 

I attempted to pull most of the composing logic out into it's own `mixin`, so it would be easier to test + read. 

[before](https://user-images.githubusercontent.com/48811365/170098557-6cd3e650-0a52-4ced-9681-54856795fbbc.mov): 

- note that composing range never updates - it stays -1, -1 regardless of composing.

[after](https://user-images.githubusercontent.com/48811365/170098681-ca1a4d72-1325-49ad-a5db-6212f7eef17c.mov):

- composing range is -1, -1 when not composing.
- composing range is the correct range when composing, including in the middle of text.

*List which issues are fixed by this PR. You must list at least one issue.*

- Fixes https://github.com/flutter/flutter/issues/65357
- Fixes https://github.com/flutter/flutter/issues/103645

<details>
<summary>
<h3>Demo Code</h3>
</summary>

```dart
void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      home: Scaffold(
        appBar: AppBar(
          title: Text('Material App Bar'),
        ),
        body: CustomInput(),
      ),
    );
  }
}

class CustomInput extends StatefulWidget {
  @override
  _CustomInputState createState() => _CustomInputState();
}

class _CustomInputState extends State<CustomInput>
    with AutomaticKeepAliveClientMixin {
  TextInputConnection? _textInputConnection;
  FocusNode? _focusNode;
  FocusAttachment? _focusAttachment;
  InputConnectionController? _input;

  int composingStart = -1;
  int composingEnd = -1;
  String text = '';

  void requestKeyboard() {
    if (_focusNode!.hasFocus) {
      _input!.openConnection(Brightness.dark);
    } else {
      FocusScope.of(context).requestFocus(_focusNode);
    }
  }

  void focusOrUnfocusIfNeeded() {
    if (_focusNode!.hasFocus) {
      _focusNode!.unfocus();
    }
  }

  @override
  bool get wantKeepAlive => _focusNode!.hasFocus;

  void _handleFocusChange() {
    if (_focusNode!.hasFocus && _focusNode!.consumeKeyboardToken()) {
      _input!.openConnection(Brightness.dark);
    } else if (!_focusNode!.hasFocus) {
      _input!.closeConnection();
    }
    updateKeepAlive();
  }

  bool get _hasFocus => _focusNode!.hasFocus;

  bool get _hasInputConnection =>
      _textInputConnection != null && _textInputConnection!.attached;

  @override
  void initState() {
    _focusNode = FocusNode();
    super.initState();
    _focusAttachment = _focusNode!.attach(context);
    _input = InputConnectionController(
        onUpdatedEditingValue: (start, end, inputtedText) {
      setState(() {
        composingEnd = end;
        composingStart = start;
        text = inputtedText;
      });
    });
    _focusNode!.addListener(_handleFocusChange);
  }

  @override
  void dispose() {
    _focusAttachment!.detach();
    _focusNode!.removeListener(_handleFocusChange);
    _input!.closeConnection();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    _focusAttachment!.reparent();
    super.build(context); // See AutomaticKeepAliveState.

    return GestureDetector(
      onTap: () => _focusNode!.requestFocus(),
      child: Column(
        children: [
          Container(
            width: 200,
            height: 200,
            color: Colors.red,
          ),
          Text('composing range: $composingStart, $composingEnd'),
          Text(text)
        ],
      ),
    );
  }
}

class InputConnectionController extends TextInputClient {
  void Function(int from, int to, String value) onUpdatedEditingValue;

  InputConnectionController({required this.onUpdatedEditingValue});

  TextInputConnection? _textInputConnection;

  bool get hasConnection =>
      _textInputConnection != null && _textInputConnection!.attached;

  void openOrCloseConnection(
      FocusNode focusNode, Brightness keyboardAppearance) {
    if (focusNode.hasFocus && focusNode.consumeKeyboardToken()) {
      openConnection(keyboardAppearance);
    } else if (!focusNode.hasFocus) {
      closeConnection();
    }
  }

  void openConnection(Brightness keyboardAppearance) {
    if (!hasConnection) {
      _textInputConnection = TextInput.attach(
        this,
        TextInputConfiguration(
          inputType: TextInputType.multiline,
          obscureText: false,
          autocorrect: false,
          inputAction: TextInputAction.newline,
          keyboardAppearance: keyboardAppearance,
          textCapitalization: TextCapitalization.sentences,
        ),
      );
    }
    _textInputConnection!.show();
  }

  void closeConnection() {
    if (hasConnection) {
      _textInputConnection!.close();
      _textInputConnection = null;
    }
  }

  @override
  void connectionClosed() {
    _textInputConnection!.connectionClosedReceived();
    _textInputConnection = null;
  }

  @override
  AutofillScope? get currentAutofillScope => null;

  @override
  TextEditingValue? get currentTextEditingValue => null;

  @override
  void performAction(TextInputAction action) {}

  @override
  void updateEditingValue(TextEditingValue value) {
    onUpdatedEditingValue(
        value.composing.start, value.composing.end, value.text);
  }

  @override
  void performPrivateCommand(String action, Map<String, dynamic> data) {}

  @override
  void insertTextPlaceholder(Size size) {
    // TODO: implement insertTextPlaceholder
  }

  @override
  void removeTextPlaceholder() {
    // TODO: implement removeTextPlaceholder
  }

  @override
  void showToolbar() {
    // TODO: implement showToolbar
  }

  @override
  void showAutocorrectionPromptRect(int start, int end) {}

  @override
  void updateFloatingCursor(RawFloatingCursorPoint point) {
    // TODO: implement updateFloatingCursor
  }
}
```
</details>

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart [hunk 1]
```diff
@@ -1295,12 +1339,6 @@ abstract class DefaultTextEditingStrategy implements TextEditingStrategy {
     }
   }
 
-  void handleCompositionUpdate(html.Event event) {
-    final EditingState newEditingState = EditingState.fromDomElement(activeDomElement);
-    editingDeltaState.composingOffset = newEditingState.baseOffset!;
-    editingDeltaState.composingExtent = newEditingState.extentOffset!;
-  }
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1302]: I honestly don't know how delta state works with composing range. Your change here seems to break delta state because it won't be able to take composing range into account anymore:
https://github.com/flutter/engine/blob/593140b9ce4207fec890c22e4b382ff92e9b389d/lib/web_ui/lib/src/engine/text_editing/text_editing.dart#L506-L512

I'd love to get @Renzo-Olivares's opinion on this one.

**Review by mdebbar — COMMENTED** [review]
Other than my concern about delta state, this PR looks good!

Thanks for taking the time to fix this!

File: lib/web_ui/lib/src/engine/text_editing/composition_aware_mixin.dart [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+///
+/// A class with this mixin should call [addCompositionEventHandlers] on initalization, and
+/// [removeCompositionEventHandlers] on deinitalization.
+///
+/// see also:
+///
+/// * [EditingState], the state of a text field that [CompositionAwareMixin] updates.
+/// * [DefaultTextEditingStrategy], the primary implementer of [CompositionAwareMixin].
+mixin CompositionAwareMixin {
+  /// the name of the HTML composition event type that triggers on starting a composition.
```
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/composition_aware_mixin.dart:24]: Start these comments with capital letters here and below.

File: lib/web_ui/lib/src/engine/text_editing/composition_aware_mixin.dart [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  static const String kCompositionStart = 'compositionstart';
```
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/composition_aware_mixin.dart:25]: Could these constants be private?  Or it looks like they're used in tests, so maybe they should be `@visibleForTesting` or even just duplicated in the tests.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart [hunk 4]
```diff
@@ -674,14 +681,20 @@ class EditingState {
   /// valid selection range for input DOM elements.
   factory EditingState.fromFrameworkMessage(
       Map<String, dynamic> flutterEditingState) {
+    final String? text = flutterEditingState.tryString('text');
+
     final int selectionBase = flutterEditingState.readInt('selectionBase');
     final int selectionExtent = flutterEditingState.readInt('selectionExtent');
-    final String? text = flutterEditingState.tryString('text');
```
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:679]: Why was this moved?

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart [hunk 5]
```diff
@@ -726,11 +757,18 @@ class EditingState {
   /// The offset at which the text selection terminates.
   final int? extentOffset;
 
+  /// The offset at which text is still being composed.
```
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:760]: What do you mean by "still being composed"?

File: lib/web_ui/test/composition_test.dart [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+            MockWithCompositionAwareMixin();
+        mockWithCompositionAwareMixin.composingText = composingText;
+
+        const int expectedComposingBase = baseOffset - composingText.length;
+
+        expect(
+            mockWithCompositionAwareMixin.determineCompositionState(editingState),
+            editingState.copyWith(
+                composingBaseOffset: expectedComposingBase,
+                composingExtentOffset: expectedComposingBase + composingText.length));
```
Comment (justinmc) [inline lib/web_ui/test/composition_test.dart:114]: Nit: I think the formatting here is wrong per the style guide, should be something like:

```dart
    composingExtentOffset: expectedComposingBase + composingText.length,
  ),
);
```

And a few other instances of this below.

File: lib/web_ui/test/composition_test.dart [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+  });
+
+  group('composing range', () {
+    test('should be [0, compostionStrLength] on new composition', () {
+      const String composingText = 'hi';
+
+      inputElement.dispatchEvent(
+          html.CompositionEvent(CompositionAwareMixin.kCompositionUpdate, data: composingText));
+
+      //set the selection text
```
Comment (justinmc) [inline lib/web_ui/test/composition_test.dart:126]: Space after the `//`, capital letter at the start and period at the end, here and a few other comments below.

**Review by justinmc — COMMENTED** [review]
Just small comments below, overall this looks good.  +1 to having @Renzo-Olivares take a look as it pertains to deltas.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart (hunk 1 above)
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1302]: It seems to be causing some issues, but i'm still looking into it. Ideally we could just pull the `composingOffset` and `composingExtent` from the `EditingState` after `determineCompositionState()` is called, but I'm getting some weird behavior doing that. I'll keep investigating, and post an update asap.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart (hunk 4 above)
Comment (antholeole) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:679]: Talked offline; it's easier to see that selectionBase and selectionExtent are grouped and composingBase and composingExtent are grouped if text is isolated. No change in functionality, just readability

File: lib/web_ui/lib/src/engine/text_editing/composition_aware_mixin.dart (hunk 3 above)
Comment (antholeole) [reply lib/web_ui/lib/src/engine/text_editing/composition_aware_mixin.dart]: made private & duplicated in tests. The name should never change anyway since it's defined in the HTML spec

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart (hunk 1 above)
Comment (Renzo-Olivares) [reply lib/web_ui/lib/src/engine/text_editing/text_editing.dart:1302]: We were previously setting the `deltaStart` and `deltaEnd` to the composing region captured in `compositionupdate`. I previously derived the composing region by grabbing the editing state from the DOM at the time when the `compositionupdate` event was fired and using the selection as the composing region. At the time of firing the `compositionupdate` event the editing states selection contains the range was replaced during composition. Previously this worked fine because we were replacing the replaced range with the composed text. However, this was a bit inaccurate since it is not the composing region after the edit was made. When making the composing region accurate through this change, a failure occurs in the _replace method because the range being replaced did not exist yet. For example when composing:

```
Type: ’n’
oldText = ‘’
deltaText = ’n’
deltaStart = ‘0’ (from composingOffset)
deltaEnd = ‘1’ (from composingExtent)
```

`_replace` tries to replace `(0,1)` with ’n’ in oldText which is an empty string. This fails and a delta is not produced. The reason for this being, by deriving the composing region from `compositionupdate` using `CompositionAwareMixin` we are actually deriving the composing region after that composing step has finished, i.e. after ’n’ has already been added to the `EditingState` the composing region is (0,1).

To fix this the `deltaStart` and `deltaEnd` should be the range replaced by the composing region, and not the composing region after the `compositionupdate` event has updated the editing state. We can do this by setting the `deltaStart` to the `composingOffset` captured by the `CompositionAwareMixin` and leaving the `deltaEnd` as the one captured by `beforeInput`. This will be the position of the caret before the `compositionupdate` event is fired. In the case of our example above it would be (0). So coming back to our previous example we would have: 

```
Type: ’n’
oldText = ‘’
deltaText = ’n’
deltaStart = ‘0’ (from composingOffset)
deltaEnd = ‘0’ (from beforeInput extentOffset)
```

Which would not cause a failure on `_replace` and generate the correct delta.

The fix would look something like https://github.com/Renzo-Olivares/engine/commit/f459888fb8275db14a8bb67a32973a65bc24bf5d . 

I'm unsure if we want to rely on 'beforeInput' or if we want to rely on 'CompositionAwareMixin' to save the selection range from the `active DOM` editing state in `compositionupdate` and set the composing region and `deltaStart` and `deltaEnd` through something like `determineCompositionStateForDelta()`.

File: lib/web_ui/test/composition_test.dart [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+          deltaModel: true,
+          onChange: (_, TextEditingDeltaState? deltaState) => deltaStream.add(deltaState)
+        );
+    });
+
+    tearDown(() {
+      editingStrategy.disable();
+    });
+
+    test('should have newly entered composing characters', () async {
```
Comment (antholeole) [inline lib/web_ui/test/composition_test.dart:207]: @Renzo-Olivares This and the test below it are the tests for your delta fix - lmk your input

Comment (Renzo-Olivares) [reply lib/web_ui/test/composition_test.dart:207]: LGTM thanks for making these

**Comment by antholeole** [comment]
Unfortunately, the failing test is not a flake - I'm able to reproduce the behavior on Firefox. Here's a video of what is happening on Firefox:

https://user-images.githubusercontent.com/48811365/171539083-073865cc-8784-486a-9b92-4bdac1bb7bb2.mov

essentially, it looks like on composing using delta mode on Firefox, a second action is required to "flush" the composed region. Not sure why this is, but it's consistent with the tests: the test failing on Firefox passes if I send a dummy event after composition. 

This is looking like a much more involved fix than originally planned because of deltas - I can continue trying to monkey patch this, but not sure if it will be beneficial, since Renzo said that deltas are getting refactored soon. 

Thoughts?

**Comment by mdebbar** [comment]
@antholeole thanks again for being patient here.

When I spoke to Renzo last week, we agreed that the refactoring was a good idea. I suggest waiting for that refactor. It should make this PR a lot simpler. What do you think @Renzo-Olivares ?

**Comment by Renzo-Olivares** [comment]
> @antholeole thanks again for being patient here.
> 
> When I spoke to Renzo last week, we agreed that the refactoring was a good idea. I suggest waiting for that refactor. It should make this PR a lot simpler. What do you think @Renzo-Olivares ?

We talked offline, but I think this can be merged since it is reproducible on master. The source of the issue is a difference in events given by Chrome and Firefox https://github.com/flutter/flutter/issues/105244#issuecomment-1145220938

**Review by mdebbar — APPROVED** [review]
LGTM

File: lib/web_ui/lib/src/engine/text_editing/composition_aware_mixin.dart [hunk 9]
```diff
@@ -0,0 +1,93 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:html' as html;
+
+import 'text_editing.dart';
+
+/// Provides default functionality for listening to HTML composition events.
```
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/composition_aware_mixin.dart:9]: Nit: Should we say "browser" composition events instead of HTML?  I don't know if the events are related to HTML exactly, or what spec they come from.

Comment (justinmc) [reply lib/web_ui/lib/src/engine/text_editing/composition_aware_mixin.dart:9]: Same for below where "HTML" is repeated.

Comment (justinmc) [reply lib/web_ui/lib/src/engine/text_editing/composition_aware_mixin.dart:9]: I see it's called `html.EventListener` in the code, maybe it's fine as-is then.

File: lib/web_ui/lib/src/engine/text_editing/composition_aware_mixin.dart [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+  static const String _kCompositionUpdate = 'compositionupdate';
+
+  /// The name of the HTML composition event type that triggers on ending a composition.
+  static const String _kCompositionEnd = 'compositionend';
+
+  late final html.EventListener _compositionStartListener = _handleCompositionStart;
+  late final html.EventListener _compositionUpdateListener = _handleCompositionUpdate;
+  late final html.EventListener _compositionEndListener = _handleCompositionEnd;
+
+  /// The currently composing text in the domElement.
```
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/composition_aware_mixin.dart:36]: Nit: Backticks around `domElement` maybe.

File: lib/web_ui/lib/src/engine/text_editing/composition_aware_mixin.dart [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+  EditingState determineCompositionState(EditingState editingState) {
+    if (editingState.baseOffset == null) {
+      return editingState;
+    }
+
+    if (composingText == null) {
+      return editingState;
+    }
+
+    if (editingState.text == null) {
```
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/composition_aware_mixin.dart:78]: Nit: Maybe one big if with ors (`||`) here?

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart [hunk 12]
```diff
@@ -708,13 +722,31 @@ class EditingState {
...
+
   /// The counterpart of [EditingState.fromFrameworkMessage]. It generates a Map that
   /// can be sent to Flutter.
   // TODO(mdebbar): Should we get `selectionAffinity` and other properties from flutter's editing state?
   Map<String, dynamic> toFlutter() => <String, dynamic>{
         'text': text,
         'selectionBase': baseOffset,
         'selectionExtent': extentOffset,
+        'composingBase': composingBaseOffset ?? -1,
+        'composingExtent': composingExtentOffset ?? -1,
```
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:749]: Is there a difference between -1 and `null` here?

This is an issue that we always face, just making sure it's on your radar.

File: lib/web_ui/lib/src/engine/text_editing/text_editing.dart [hunk 13]
```diff
@@ -726,11 +758,18 @@ class EditingState {
...
+  /// The offset at which [CompositionAwareMixin.composingText] terminates, if any.
+  final int? composingExtentOffset;
+
   /// Whether the current editing state is valid or not.
   bool get isValid => baseOffset! >= 0 && extentOffset! >= 0;
 
   @override
-  int get hashCode => Object.hash(text, baseOffset, extentOffset);
+  int get hashCode => Object.hash(
+    text, baseOffset, extentOffset, composingBaseOffset, composingExtentOffset);
```
Comment (justinmc) [inline lib/web_ui/lib/src/engine/text_editing/text_editing.dart:772]: Nit: I think the close parenthesis goes on the next line according to the styleguide.

File: lib/web_ui/test/composition_test.dart [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+import 'package:ui/src/engine/text_editing/input_type.dart';
+import 'package:ui/src/engine/text_editing/text_editing.dart';
+
+void main() {
+  internalBootstrapBrowserTest(() => testMain);
+}
+
+class _MockWithCompositionAwareMixin with CompositionAwareMixin {
+  // These variables should be equal to their counterparts in CompositionAwareMixin.
+  // Seperate so the counterparts in CompositionAwareMixin can be private.
```
Comment (justinmc) [inline lib/web_ui/test/composition_test.dart:23]: "Seperate" => "Separate"

File: lib/web_ui/test/composition_test.dart [hunk 15]
```diff
@@ -0,0 +1,258 @@
...
+  static const String _kCompositionEnd = 'compositionend';
+}
+
+html.InputElement get _inputElement {
+  return defaultTextEditingRoot.querySelectorAll('input').first as html.InputElement;
+}
+
+GloballyPositionedTextEditingStrategy _enableEditingStrategy({
+  required bool deltaModel,
+  void Function(EditingState?, TextEditingDeltaState?)? onChange,
```
Comment (justinmc) [inline lib/web_ui/test/composition_test.dart:35]: I think these 2 lines should be indented.

File: lib/web_ui/test/composition_test.dart [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+  return editingStrategy;
+}
+
+Future<void> testMain() async {
+  await initializeEngine();
+
+  const String fakeComposingText = 'ImComposingText';
+
+  group('$CompositionAwareMixin', () {
+    late TextEditingStrategy editingStrategy;
```
Comment (justinmc) [inline lib/web_ui/test/composition_test.dart:56]: Can this be late final?

File: lib/web_ui/test/composition_test.dart [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+      test('should set composing text to event composing text', () {
+        const String fakeEventText = 'IAmComposingThis';
+        final _MockWithCompositionAwareMixin mockWithCompositionAwareMixin =
+            _MockWithCompositionAwareMixin();
+        mockWithCompositionAwareMixin.composingText = fakeComposingText;
+        mockWithCompositionAwareMixin.addCompositionEventHandlers(_inputElement);
+
+        _inputElement.dispatchEvent(html.CompositionEvent(
+            _MockWithCompositionAwareMixin._kCompositionUpdate,
+            data: fakeEventText));
```
Comment (justinmc) [inline lib/web_ui/test/composition_test.dart:102]: Another place where I think the close parentheses should go on the next line.

File: lib/web_ui/test/composition_test.dart [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+          isA<EditingState>()
+              .having((EditingState editingState) => editingState.composingBaseOffset!,
+                  'composingBaseOffset', beforeComposingText.length - composingText.length)
+              .having((EditingState editingState) => editingState.composingExtentOffset,
+                  'composingExtentOffset', beforeComposingText.length));
+    });
+  });
+
+  group('Text Editing Delta Model', () {
+    late GloballyPositionedTextEditingStrategy editingStrategy;
```
Comment (justinmc) [inline lib/web_ui/test/composition_test.dart:192]: Another place where maybe you can use `late final`.

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Just a bunch of small comments below.

**Comment by antholeole** [comment]
🤞
